### DISCUSSION TITLE
I2C not working

### DISCUSSION THREAD

**Discussion by rbianchi99** [micropython/micropython#16078] [STM32 / Pyboard]
Hi, I am stuck. 
I'm working with a Nucleo F401RE board from STM. I use Windows.
I correctly installed micropython on the board, and it works fine, connecting to the repl with mpremote.
I'm trying to connect a Sensirion SDP33 to the board, but when I go and try to initialize a I2C connection, it doesn't work.
I have connected the power supply cables to the 3v3 and GND on the board, in the Arduino connectors.

Then i have connected to the PB6 and PB7 pins for the SCL and SDA to the corresponding cables, since STM32X shows them as default pins for the first I2C line.

I haven't programmed the board to have them as set for I2C using STM32X and STMCubeIDE, because if I do that and than try and add the hex file to the board, Micropython stops working. It still boots and gets me to the REPL, but it doesn't let me write anything, it just stays stuck there.

I'm not using any additional software, like Teraterm or Putty, because I need for this to work using only micropython, so i use mprmemote, which works fine if i don't try and fuck up anything.

So, i tried just using micropython and putting the following code in (Each command is of course on its own line, idk how to break line on this website, I'm very much inexperienced) :

`from machine import Pin
from machine import I2C
i2c=I2C(scl=Pin('PB6'), sda=Pin('PB7') )
i2c.scan()`

I get [], which means that the connection is not estabilished. 

I checked the cables and they work. I can't check the power supply from the arduino, but the pin on the same line gives out the voltage just fine, so it should be ok.

I have no idea of what I'm failing at.

**Comment by peterhinch** [comment]
I2C requires pullup resistors on both lines. It's possible that the board has these, but if not you need to provide them. Between 1KΩ and 10KΩ, to 3.3V.

Re posting code, please read [this](https://github.com/orgs/micropython/discussions/9111).

**Reply by rbianchi99** [reply]
Thank you for your reply.
I checked and in the schematics I see that indeed there are 4.7KΩ resistors between the 3.3v source and the scl/sda line of the sensor. 
I also checked if the current is going to the board and indeed it is. 
Also I double checked if the pins activated or not when initiating the I2C connection and they indeed turn on. 
I still have no idea of what's wrong.

**Reply by scruss** [reply]
Please show us a picture of your wiring. Your mention of "the power supply from the arduino" makes me concerned about how the sensor is getting power

**Comment by rbianchi99** [comment]
![GG](https://github.com/user-attachments/assets/fbfdb803-77a7-4d32-af25-3ada172745f7)

Here you go. 
Extremely professional wiring, as you can see, lol
Thank you for your help

**Reply by rkompass** [reply]
Another trivial cause might be that the pins are not those they appear to be. This you may easily check by setting the pins to output mode and changing their value followed by measuring their voltage manually (with voltage meter or with an LED in series with a resistor of, say 200 Ohms).
E.g.
```
from machine import Pin
pb6 = Pin('PB6', mode=Pin.OUT)
pb6(1)
# measure
pb6(0)
# measure
```

**Reply by rkompass** [reply]
Also you may try 
`i2c = SoftI2C(.....)`
which I have found to work sometimes more reliably imstead of I2C.
Specifying a low frequency in the I2C constructor may also help.
Finally I would try to establish an I2C connection with an I2C module that you found to function already in a different setup (to exclude that your SDP33 is broken).

**Comment by sosi-deadeye** [comment]
Usually microcontrollers could map GPIOs to different Hardware-Implementations like `I2C`, `SPI` etc.
Try to use the dedicated GPIOs for `I2C`.
![sda_scl](https://github.com/user-attachments/assets/86881fba-c574-44f1-826b-c6168519b3bd)
